### PR TITLE
Added state tests 

### PR DESCRIPTION
**Pull request by efe-ozbatur**
Added new state tests checking for emergencies and other conditions

### CODE REVIEW

**Comment by mifrandir** [comment]
Everyone make sure that in the Randomiser the data that's being generated makes sense. In particular, make sure that we are not leaving out any critical or sensible values.

@Videosquared @hur @suryanshmanocha @aleksanderT23

File: test/src/state_machine/state.test.cpp [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+
+    if (!has_emergency && !calibrate_command) {
+      bool all_initialised = checkModulesInitialised(log, embrakes_data, nav_data, batteries_data,
+                                                     telemetry_data, sensors_data, motors_data);
+      hyped::state_machine::State *new_state = state->checkTransition(log);
+
+      if (all_initialised) {
+        ASSERT_EQ(new_state, FailureStopped::getInstance());
+      } else {
+        ASSERT_NE(new_state, FailureStopped::getInstance());
```
Comment (mifrandir) [inline test/src/state_machine/state.test.cpp]: This seems wrong to me. Surely we should make sure that in the case where there is no emergency, we have received the calibrating command and all modules are initialised we go into Calibrating.

File: test/src/state_machine/state.test.cpp [hunk 2]
```diff
@@ -0,0 +1,781 @@
...
+    data.setMotorData(motors_data);
+
+    bool has_emergency = checkEmergency(log, embrakes_data, nav_data, batteries_data,
+                                        telemetry_data, sensors_data, motors_data);
+
+    if (!has_emergency) {
+      bool stopped                           = checkPodStopped(log, nav_data);
+      hyped::state_machine::State *new_state = state->checkTransition(log);
+
+      if (stopped) { ASSERT_EQ(new_state, Finished::getInstance()); }
```
Comment (mifrandir) [inline test/src/state_machine/state.test.cpp]: I think it's possible to add the `else` clause here.

**Review by mifrandir — CHANGES_REQUESTED** [review]

File: test/src/state_machine/state.test.cpp [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+  for (int i = 0; i < TEST_SIZE; i++) {
+    Randomiser::randomiseEmbrakes(embrakes_data);
+    Randomiser::randomiseNavigation(nav_data);
+    Randomiser::randomiseTelemetry(telemetry_data);
+    Randomiser::randomiseMotors(motors_data);
+
+    data.setEmergencyBrakesData(embrakes_data);
+    data.setNavigationData(nav_data);
+    data.setTelemetryData(telemetry_data);
+    data.setMotorData(motors_data);
```
Comment (mifrandir) [inline test/src/state_machine/state.test.cpp:308]: Here (and in other tests) we are only randomising 4 modules but we're actually using 6. Why is that?

What's up with batteries and sensors?

**Review by mifrandir — CHANGES_REQUESTED** [review]

File: test/src/state_machine/state.test.cpp [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+
+    bool has_emergency = checkEmergency(log, embrakes_data, nav_data, batteries_data,
+                                        telemetry_data, sensors_data, motors_data);
+    hyped::state_machine::State *new_state = state->checkTransition(log);
+
+    if (has_emergency) {
+      ASSERT_EQ(new_state, FailureStopped::getInstance());
+    } else {
+      ASSERT_NE(new_state, FailureStopped::getInstance());
```
Comment (pmorande27) [inline test/src/state_machine/state.test.cpp]: Need to add error messages in case that the assert fails. Looking good :)

**Review by pmorande27 — CHANGES_REQUESTED** [review]

File: test/src/state_machine/state.test.cpp [hunk 5]
```diff
@@ -0,0 +1,781 @@
...
+    dup2(stdout_f, 1);
+    close(stdout_f);
+  }
+};
+
+//---------------------------------------------------------------------------
+// Test structures
+//---------------------------------------------------------------------------
+
+struct IdleTest : public StateTest {
```
Comment (DonaldJennings) [inline test/src/state_machine/state.test.cpp]: Try and populate these structs if possible. Additionally, for better readability in the test files, these structs should be above the suite of tests that use them. If something is specific to a certain test suite and is repeated throughout tests then add it here :)

File: test/src/state_machine/state.test.cpp [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+//--------------------------------- TESTS -----------------------------------
+//---------------------------------------------------------------------------
+
+//---------------------------------------------------------------------------
+// Idle Tests
+//---------------------------------------------------------------------------
+
+TEST_F(IdleTest, handlesEmergency)
+{
+  Data &data  = Data::getInstance();
```
Comment (DonaldJennings) [inline test/src/state_machine/state.test.cpp]: could we possibly find a way of extracting these variables into the test structs?

File: test/src/state_machine/state.test.cpp [hunk 7]
```diff
@@ -0,0 +1,781 @@
+/*
+ * Author: Efe Ozbatur, Franz Miltz
+ * Organisation: HYPED
+ * Date: 06.04.2021
+ *
+ *    Copyright 2019 HYPED
```
Comment (DonaldJennings) [inline test/src/state_machine/state.test.cpp]: The Copyright here should be updated to 2021

File: test/src/state_machine/state.test.cpp [hunk 8]
```diff
@@ -0,0 +1,781 @@
...
+struct FailureBrakingTest : public StateTest {
+};
+
+struct FailureStoppedTest : public StateTest {
+};
+
+//---------------------------------------------------------------------------
+// Randomiser
+//---------------------------------------------------------------------------
+
```
Comment (DonaldJennings) [inline test/src/state_machine/state.test.cpp:38]: Perhaps add a class comment here explaining more about what the Randomiser is doing

**Review by DonaldJennings — CHANGES_REQUESTED** [review]

File: test/src/state_machine/state.test.cpp [hunk 9]
```diff
@@ -275,17 +319,18 @@ class Randomiser {
...
 
+
 //---------------------------------------------------------------------------
 // Idle Tests
 //---------------------------------------------------------------------------
 
-TEST_F(IdleTest, handlesEmergency)
-{
-  Data &data  = Data::getInstance();
+struct IdleTest : public StateTest {
```
Comment (pmorande27) [inline test/src/state_machine/state.test.cpp]: I would add a comment here explaining what this set of tests is going to test

**Review by pmorande27 — COMMENTED** [review]
I think that everything is ok. I would just add some comments to each of the structs :)

**Review by pmorande27 — APPROVED** [review]
All ok :)

**Review by mifrandir — APPROVED** [review]

**Review by CallumAlexander — APPROVED** [review]
All good.

**Review by DonaldJennings — APPROVED** [review]

File: test/src/state_machine/state.test.cpp [hunk 10]
```diff
@@ -0,0 +1,771 @@
...
+
+  //---------------------------------------------------------------------------
+  // Motor data
+  //---------------------------------------------------------------------------
+
+  static void randomiseMotors(Motors &motors_data)
+  {
+    randomiseModuleStatus(motors_data.module_status);
+
+    // Generates a RPM data between 0 and 199 for all 4 motors.
```
Comment (aleksanderT23) [inline test/src/state_machine/state.test.cpp]: According to the calculations checked by propulsion, the range of rpm's should be between 33 and 1400, so the code entry is (rand() % 1368 + 33);

**Review by aleksanderT23 — CHANGES_REQUESTED** [review]
According to the calculations checked by propulsion, the range of rpm's should be between 33 and 1400, so the code entry is (rand() % 1368 + 33);

    for (int i = 0; i < motors_data.kNumMotors; i++) {
      motors_data.rpms[i] = static_cast<uint32_t>(rand() % 1368 + 33);
    }
  }

File: test/src/state_machine/state.test.cpp (hunk 10 above)
Comment (mifrandir) [reply test/src/state_machine/state.test.cpp]: I've now set it to [0, 2000] which includes the given range. Cheers

**Review by aleksanderT23 — APPROVED** [review]
Seems good!

**Pull request merged**